drivers/bluetooth/hci/open_amp_nrf.c

```diff
@@ -0,0 +1,465 @@
+
+	/* Libmetal setup */
+	struct metal_init_params metal_params = METAL_INIT_DEFAULTS;
+	status = metal_init(&metal_params);
+	if (status != 0) {
+		BT_ERR("metal_init: failed - error code %d", status);
+		return status;
+	}
+
+	status = metal_register_generic_device(&shm_device);
+	if (status != 0) {
+		BT_ERR("Couldn't register shared memory device: %d", status);
+		return status;
+	}
+
+	status = metal_device_open("generic", SHM_DEVICE_NAME, &device);
+	if (status != 0) {
+		BT_ERR("metal_device_open failed: %d", status);
+		return status;
+	}
+
+	io = metal_device_io_region(device, 0);
+	if (!io) {
+		BT_ERR("metal_device_io_region failed to get region");
+		return -ENODEV;
+	}
+
+	/* IPM setup */
+	ipm_tx_handle = device_get_binding("IPM_0");
+	if (!ipm_tx_handle) {
+		BT_ERR("Could not get TX IPM device handle");
+		return -ENODEV;
+	}
+
+	ipm_rx_handle = device_get_binding("IPM_1");
+	if (!ipm_rx_handle) {
+		BT_ERR("Could not get RX IPM device handle");
+		return -ENODEV;
+	}
+
+	ipm_register_callback(ipm_rx_handle, ipm_callback, NULL);
+
+	/* Virtqueue setup */
+	vq[0] = virtqueue_allocate(VRING_SIZE);
+	if (vq[0] == NULL) {
+		BT_ERR("virtqueue_allocate failed to alloc vq[0]");
+		return -ENOMEM;
+	}
+
+	vq[1] = virtqueue_allocate(VRING_SIZE);
+	if (vq[1] == NULL) {
+		BT_ERR("virtqueue_allocate failed to alloc vq[1]");
+		return -ENOMEM;
+	}
+
+	rvrings[0].io = io;
+	rvrings[0].info.vaddr = (void *)VRING_TX_ADDRESS;
+	rvrings[0].info.num_descs = VRING_SIZE;
+	rvrings[0].info.align = VRING_ALIGNMENT;
+	rvrings[0].vq = vq[0];
+
+	rvrings[1].io = io;
+	rvrings[1].info.vaddr = (void *)VRING_RX_ADDRESS;
+	rvrings[1].info.num_descs = VRING_SIZE;
+	rvrings[1].info.align = VRING_ALIGNMENT;
+	rvrings[1].vq = vq[1];
+
+	vdev.role = RPMSG_MASTER;
+	vdev.vrings_num = VRING_COUNT;
+	vdev.func = &dispatch;
+	vdev.vrings_info = &rvrings[0];
+
+	rpmsg_virtio_init_shm_pool(&shpool, (void *)SHM_START_ADDR, SHM_SIZE);
+	status = rpmsg_init_vdev(&rvdev, &vdev, ns_bind_cb, io, &shpool);
+	if (status != 0) {
+		BT_ERR("rpmsg_init_vdev failed %d", status);
+		return status;
+	}
+
+	/* Since we are using name service, we need to wait for a response
+	* from NS setup and than we need to process it
+	*/
+	virtqueue_notification(vq[0]);
+
+	/* Wait til nameservice ep is setup */
+	k_sem_take(&sync_sem, K_FOREVER);
+
+	return 0;
+}
+
+static inline void bt_buf_set_prio(struct net_buf *buf, bool prio)
+{
+	*((u8_t *)net_buf_user_data(buf) + 1) = prio;
+}
+
+static inline bool bt_buf_get_prio(struct net_buf *buf)
+{
+	return (bool)(*((u8_t *)net_buf_user_data(buf) + 1));
+}
+
+static void rx_thread(void *p1, void *p2, void *p3)
+{
+	struct net_buf *buf;
+
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	BT_DBG("Rx thread started");
+
+	while (1) {
+		buf = net_buf_get(&rx_queue, K_FOREVER);
+		do {
+			BT_DBG("Calling bt_recv(%p)", buf);
+			if (bt_buf_get_prio(buf)) {
+				bt_recv_prio(buf);
+			} else {
+				bt_recv(buf);
+			}
```

Comment:
The whole idea of the `bt_recv_prio()` vs. `bt_recv()` separation is that they do *not* happen from the same context (thread). This is particularly important when you're implementing the RX thread on the HCI driver side rather than letting the host do it. In this case `bt_recv_prio()` should probably be used by whatever context is feeding `rx_queue` and here in the rx_thread you'd only use `bt_recv()`

Reply:
Actually, you're not selecting BT_RECV_IS_RX_THREAD in your Kconfig definition but still creating your own RX thread inside the driver? You need to decide whether you want to keep the thread inside the driver or the host (or perhaps both, which would have a substantial memory overhead). As long as you don't select BT_RECV_IS_RX_THREAD it is actually safe to call the recv() APIs like you're doing since the host-side will take care of splitting processing into independent threads, however if you did mean to select BT_RECV_IS_RX_THREAD then you need to separate them inside the driver.

Reply:
> Actually, you're not selecting BT_RECV_IS_RX_THREAD in your Kconfig definition but still creating your own RX thread inside the driver? 

That `BT_RECV_IS_RX_THREAD` option is still poorly named, but the documentation is better.
```

Reply:
Isn't that the other way around? The host has its own RX thread when `BT_RECV_IS_RX_THREAD` is *not* enabled

Reply:
> Isn't that the other way around? The host has its own RX thread when BT_RECV_IS_RX_THREAD is not enabled

Yes. The intention of the current option name is to say "The bt_recv() call happens in the Bluetooth RX thread", which then implies that there's no need for the host to have such a thread. I'm open to better names if you think it's confusing.

Reply:
> Since we call it from the same priority context, we need to select the BT_RECV_IS_RX_THREAD option.

But do you need to call it from the same context? What's the reason there's something called the RX thread inside the driver, and that thread uses the Bluetooth RX thread Kconfig options like BT_RX_THREAD_SIZE. That makes it very much look like "the" Bluetooth RX thread :smile:

So, do you need a separate thread like this in the driver? If not, just call `bt_recv()` and `bt_recv_prio()` directly from the context that currently feeds the `rx_queue` and let the host take care of the details. If you do need the separate thread, move the `bt_recv_prio()` like I suggested and select the `BT_RECV_IS_RX_THREAD` option.

Reply:
> Since we call it from the same priority context, we need to select the BT_RECV_IS_RX_THREAD option.

Like Carles said it's actually the opposite. My previous comment was based on imagining a "don't" in the above quote 😄

Reply:
> Isn't that the other way around? The host has its own RX thread when `BT_RECV_IS_RX_THREAD` is **not** enabled

Yes, sorry. Messed it up.

In h4.c bt_recv is called from thread. and bt_recv_prio is called from interrupt.
In the hci_driver.c bt_recv is called from a thread, and bt_recv_prio is called from a different thread with higher priority.
In open_amp_nrf.c bt_recv and bt_recv_prio is called from the same thread.

So h4 and hci_driver both fulfull the requirement of sending from different priority contexts, and can select BT_RECV_IS_RX_THREAD.
This one cannot select it because it it calls them from the same priority context.

So then this is correct.

Reply:
> So then this is correct.

It's not wrong from a HCI driver API contract perspective, but my question still remains: is this "RX thread" in the driver actually needed, and why should it have its stack-size determined by the Kconfig option when it never reaches the application level (that's why the normal RX thread size is user-configurable).

Reply:
Thank you for clarifying. I completely misunderstood your comment. And I think you have a good point (assuming I have understood it correctly :P).

@kapi-no Have you considered getting rid of the TX thread here and calling bt_recv and bt_recv_prio directly from `bt_open_amp_rx`. In `bt_recv` we are basically doing the same thing as what is being done here (see here: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L4866). That would mean that `bt_recv_prio` would be processed in the context of `bt_open_amp_rx`, which seem to be interrupt context.
So I guess the real question is if there is any downside to calling `bt_recv_prio` directly from interrupt context (there are other HCI drivers that do exactly that).